### PR TITLE
Introduce a mechanism to collect several errors in a module

### PR DESCRIPTION
**Pull request by josevalim**
Several operations in Elixir lead to an invalid module: missing local functions, conflicting imports, etc. This PR introduces a mechanism to report multiple errors and abort compiling modules:

Screenshot 2022-11-27 at 17 51 52

This PR is currently incomplete. TODO:

* [x] Send errors as diagnostic
* [x] List all modules to analyze

Modules to analyze:

* [x] elixir_aliases
* [ ] elixir_bitstring
* [x] elixir_clauses
* [x] elixir_dispatch 
* [x] elixir_def
* [ ] elixir_expand
* [x] elixir_map
* [x] elixir_module
* [x] elixir_overridable

### CODE REVIEW

**Comment by josevalim** [comment]
It is also important noting that there are many other compilation errors that will crash. Therefore, a downside of this change is that it introduces an additional mode of failure in Elixir code. Some of them are printed warnings, some of them are printed errors, some of them are actual exceptions.

**Comment by lukaszsamson** [comment]
Does it affect Kernel.ParallelCompiler or Mix.Diagnostic API?

**Comment by josevalim** [comment]
@lukaszsamson it will in the sense that you will have several error diagnostics at once (today it is at most one unless --warning-as-errors is true).

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

File: lib/elixir/src/elixir_errors.erl [hunk 1]
```diff
@@ -14,89 +16,91 @@
...
+  print_error(Meta, Env, Module, Desc),
+  case elixir_module:taint(EnvModule) of
+    true -> ok;
+    false -> compile_error(Env)
+  end;
+module_error(Meta, Env, Module, Desc) ->
+  file_error(Meta, Env, Module, Desc).
+
+%% A function error is one where it can continue if there is a function
+%% being compiled. If there is no function, it is falls back to module_error.
```
Comment (sabiwara) [inline lib/elixir/src/elixir_errors.erl:71]: Maybe I misunderstand, but based on the current implementation it seems to be
```diff
-%% being compiled. If there is no function, it is falls back to module_error.
+%% being compiled. If there is no function, it is falls back to file_error.
```

Should the fallback be to `module_error`?

Comment (josevalim) [reply lib/elixir/src/elixir_errors.erl:71]: Yes, good catch, I will fix.

**Pull request merged**